text_generation/causal_lm/cpp/continuous_batching/library/src/scheduler.hpp

```diff
@@ -89,54 +90,94 @@ class Scheduler {
         return num_running;
     }
 
-    void _preempt_by_recompute(SequenceGroup::Ptr sequence_group) {
-        // currently, we support only preemption by (TODO: implement "partial") recompute
+
+    bool _preempt_by_recompute(SequenceGroup::Ptr sequence_group, size_t blocks_needed) {
+        size_t total_num_released_blocks = 0;
+        size_t processed_tokens = sequence_group->get_num_processed_tokens();
+        size_t block_size = m_config.block_size;
+        size_t prev_blocks_count = m_block_manager.num_free_blocks();
+        size_t preempted_tokens = 0;
+
+        // preempt whole sequence group for non-grady sampling
+        if (!sequence_group->get_sampling_parameters().is_gready_sampling()) {
```

Comment:
maybe we can change condition to `num_running_sequences > 1`?

Reply:
Done,